### PR TITLE
IMPRO-1330: Cap alpha values

### PR DESCRIPTION
**Pull request by neilCrosswaite**
Capped the alphas values at 0.5.
given warnings at any more iterations than 2.
changed the cubes and bats to be acceptable and therefore changed the KGO to allow for this.
Tested for the errors and warning.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

File: lib/improver/nbhood/recursive_filter.py [hunk 1]
```diff
@@ -73,29 +79,35 @@ def __init__(self, alpha_x=None, alpha_y=None, iterations=None,
...
-        if alpha_y is not None:
-            if not 0 < alpha_y < 1:
-                raise ValueError(
-                    "Invalid alpha_y: must be > 0 and < 1: {}".format(
-                        alpha_y))
 
+        """
+        alpha_error = ("alpha must be less than 0.5. A large alpha value"
+                       "leads to poor conservation of probabilities: ")
+        if alpha_x is not None and not 0 < alpha_x <= 0.5:
```
Comment (arh89) [inline lib/improver/nbhood/recursive_filter.py]: (We could ~halve the number of lines of code here by doing this in a loop)

File: lib/improver/nbhood/recursive_filter.py [hunk 2]
```diff
@@ -73,29 +79,35 @@ def __init__(self, alpha_x=None, alpha_y=None, iterations=None,
...
+            raise ValueError(message.format(alpha_y))
         if iterations is not None:
-            if not iterations >= 1:
+            if iterations < 1:
                 raise ValueError(
                     "Invalid number of iterations: must be >= 1: {}".format(
                         iterations))
+            if iterations > 2:
+                warnings.warn(
+                    "Iterations of two will lead to poorer conservation.")
```
Comment (arh89) [inline lib/improver/nbhood/recursive_filter.py]: `iterations > 2`, yet the message implies `iterations == 2` - additionally, is this message actually useful to an ed user, or should it be obvious from the algorithm? What if `iterations == 3`, is it poor then also?

File: lib/improver/nbhood/recursive_filter.py [hunk 3]
```diff
@@ -319,7 +331,21 @@ def process(self, cube, alphas_x=None, alphas_y=None, mask_cube=None):
             new_cube (iris.cube.Cube):
                 Cube containing the smoothed field after the recursive filter
                 method has been applied.
+
+        Raises:
+            ValueError: If any alpha cube value is over 0.5
         """
+
```
Comment (arh89) [inline lib/improver/nbhood/recursive_filter.py]: (Similarly, we can halve the number of lines of code here by using a loop)

File: lib/improver/nbhood/recursive_filter.py (hunk 2 above)
Comment (MoseleyS) [reply lib/improver/nbhood/recursive_filter.py]: The ticket asks for a warning if more than 2 iterations are used, so it is the message that needs updating here. How about "More than two iterations degrades the conservation of probability assumption."

File: lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py [hunk 4]
```diff
@@ -95,35 +96,31 @@ class Test__init__(Test_RecursiveFilter):
 
     def test_alpha_x_gt_unity(self):
         """Test when an alpha_x value > unity is given (invalid)."""
```
Comment (MoseleyS) [inline lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py]: It isn't greater than unity any more (update doc-string)

File: lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py [hunk 5]
```diff
@@ -95,35 +96,31 @@ class Test__init__(Test_RecursiveFilter):
...
         alpha_x = -0.5
-        msg = "Invalid alpha_x: must be > 0 and < 1: -0.5"
+        msg = "Invalid alpha_x: must be > 0 and <= 0.5: -0.5"
         with self.assertRaisesRegex(ValueError, msg):
-            RecursiveFilter(alpha_x=alpha_x, alpha_y=None,
-                            iterations=None, edge_width=1)
+            RecursiveFilter(alpha_x=alpha_x)
 
     def test_alpha_y_gt_unity(self):
         """Test when an alpha_y value > unity is given (invalid)."""
```
Comment (MoseleyS) [inline lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py]: Update docstring

File: lib/improver/cli/recursive_filter.py [hunk 6]
```diff
@@ -135,7 +135,9 @@ def process(cube, mask_cube=None, alphas_x_cube=None, alphas_y_cube=None,
             square in the y direction.
             Default is None.
         iterations (int):
-            Number of times to apply the filter. (Typically < 5)
+            Number of times to apply the filter. (Typically < 3)
+            Number of iterations should be less than 3, any high and it may
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/cli/recursive_filter.py]: Done, but with a new line after been.

File: lib/improver/nbhood/recursive_filter.py [hunk 7]
```diff
@@ -50,6 +51,11 @@ def __init__(self, alpha_x=None, alpha_y=None, iterations=None,
         """
         Initialise the class.
 
+        The alpha values determine how much "value" of a cell undergoing
+        filtering os comprised of the current value at that cell and how much
```
[the remark this answers is not in the document]
Comment (neilCrosswaite) [reply lib/improver/nbhood/recursive_filter.py]: done

File: lib/improver/nbhood/recursive_filter.py (hunk 1 above)
Comment (neilCrosswaite) [reply lib/improver/nbhood/recursive_filter.py]: Looped it up.

Not sure about moving it all to process, yes it is being split, but that is where the arguments are being given to the object. There is no consistent way of this being done throughout the code.

File: lib/improver/nbhood/recursive_filter.py (hunk 2 above)
Comment (neilCrosswaite) [reply lib/improver/nbhood/recursive_filter.py]: Thank you. done.

File: lib/improver/nbhood/recursive_filter.py (hunk 3 above)
Comment (neilCrosswaite) [reply lib/improver/nbhood/recursive_filter.py]: Done

File: lib/improver/nbhood/recursive_filter.py (hunk 1 above)
Comment (MoseleyS) [reply lib/improver/nbhood/recursive_filter.py]: Both `__init__` and `process` could call a function containing this bit?

File: lib/improver/cli/recursive_filter.py [hunk 8]
```diff
@@ -135,7 +135,9 @@ def process(cube, mask_cube=None, alphas_x_cube=None, alphas_y_cube=None,
             square in the y direction.
             Default is None.
         iterations (int):
-            Number of times to apply the filter. (Typically < 5)
+            Number of times to apply the filter. (Typically < 3)
+            Number of iterations should be less than 3, higher values have been
+            show to lead to poorer conservation.
```
Comment (MoseleyS) [inline lib/improver/cli/recursive_filter.py]:
```suggestion
            shown to lead to poorer conservation.
```

File: lib/improver/nbhood/recursive_filter.py [hunk 9]
```diff
@@ -319,7 +328,16 @@ def process(self, cube, alphas_x=None, alphas_y=None, mask_cube=None):
...
                 Cube containing the smoothed field after the recursive filter
                 method has been applied.
+
+        Raises:
+            ValueError: If any alpha cube value is over 0.5
         """
+        for alpha in (alphas_x, alphas_y):
+            if alpha is not None and (alpha.data > 0.5).any():
+                raise ValueError(
+                    "alpha must be less than 0.5. A large alpha value "
```
Comment (MoseleyS) [inline lib/improver/nbhood/recursive_filter.py]:
```suggestion
                    "All alpha values must be less than 0.5. A large alpha value "
```

File: lib/improver/nbhood/recursive_filter.py [hunk 10]
```diff
@@ -92,23 +92,20 @@ def __init__(self, alpha_x=None, alpha_y=None, iterations=None,
...
                        "leads to poor conservation of probabilities: ")
-        if alpha_x is not None and not 0 < alpha_x <= 0.5:
-            message = alpha_error if alpha_x > 0.5 else ''
-            message += "Invalid alpha_x: must be > 0 and <= 0.5: {}"
-            raise ValueError(message.format(alpha_x))
-        if alpha_y is not None and not 0 < alpha_y <= 0.5:
-            message = alpha_error if alpha_y > 0.5 else ''
-            message += "Invalid alpha_y: must be > 0 and <= 0.5: {}"
-            raise ValueError(message.format(alpha_y))
+        for k, v in {'x': alpha_x, 'y': alpha_y}.items():
```
Comment (arh89) [inline lib/improver/nbhood/recursive_filter.py]: Sorry, one last nitpick - can we please use better variable names than `k`, and `v` please?
e.g: `direction, alpha` or just anything slightly more specific.

Comment (neilCrosswaite) [reply lib/improver/nbhood/recursive_filter.py]: `i, j`?

just kidding.
Fixed.

Comment (arh89) [reply lib/improver/nbhood/recursive_filter.py]: Thanks

File: lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py [hunk 11]
```diff
@@ -133,6 +130,19 @@ def test_iterations(self):
             RecursiveFilter(alpha_x=None, alpha_y=None,
                             iterations=iterations, edge_width=1)
 
+    @ManageWarnings(record=True)
+    def test_interations_warn(self, warning_list=None):
```
Comment (bayliffe) [inline lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py]: iterations, not interations.

File: lib/improver/cli/recursive_filter.py [hunk 12]
```diff
@@ -135,7 +135,9 @@ def process(cube, mask_cube=None, alphas_x_cube=None, alphas_y_cube=None,
             square in the y direction.
             Default is None.
         iterations (int):
-            Number of times to apply the filter. (Typically < 5)
+            Number of times to apply the filter. (Typically < 3)
+            Number of iterations should be less than 3, higher values have been
```
Comment (neilCrosswaite) [inline lib/improver/cli/recursive_filter.py]:
```suggestion
            Number of iterations should be 2 or less, higher values have been
```

File: lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py (hunk 11 above)
Comment (neilCrosswaite) [reply lib/improver/tests/nbhood/recursive_filter/test_RecursiveFilter.py]:
```suggestion
    def test_iterations_warn(self, warning_list=None):
```

File: lib/improver/nbhood/recursive_filter.py [hunk 13]
```diff
@@ -73,30 +79,33 @@ def __init__(self, alpha_x=None, alpha_y=None, iterations=None,
...
-                        alpha_y))
 
+        """
+        alpha_error = ("alpha must be less than 0.5. A large alpha value"
+                       "leads to poor conservation of probabilities: ")
+        for k, v in {'x': alpha_x, 'y': alpha_y}.items():
+            if v is not None and not 0 < v <= 0.5:
+                message = alpha_error if v > 0.5 else ''
+                message += "Invalid alpha_{}: must be > 0 and <= 0.5: {}"
+                raise ValueError(message.format(k, v))
```
[the remark this answers is not in the document]
Comment (arh89) [reply lib/improver/nbhood/recursive_filter.py]: Could've just changed alphas to a list/tuple, and/or adjusted the error message, then we wouldn't have needed the dictionary.

**Pull request merged**